### PR TITLE
Add CustomEnityLookupSkill to 2020-06-30 and 2020-06-30-Preview

### PR DESCRIPTION
**Pull request by shuyangmsft**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json
Comment (heaths) [inline specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: 2020-06-30 already GA'd. Unless this support already existed and just wasn't formalized in swagger, it should only be in 2020-06-30-Preview, which will change dates when closer to its GA.

**Review by heaths — CHANGES_REQUESTED** [review]

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (heaths) [inline specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6982]: What's the default value?

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (heaths) [inline specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6947]: Will any of these be sent back in a response as explicit `null`? If so, they must be marked `"x-nullable": true`.

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (heaths) [inline specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6970]: You should specify the default value in swagger as well.

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (shuyangmsft) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6947]: Except "name", all other parameters are optional. Per previous discussion, we should save bandwidth to not use "x-nullable" here.

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (shuyangmsft) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6982]: There 3 default values are optional and should be provided by customers. If customers don't provide any value, it's still OK and in our backend service, the default value will just be null.

File: specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json
Comment (shuyangmsft) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: I actually have a question here given our case is a little bit involved.
The CustomEntityLookupSkill is publicly available. It's currently in our preview API and customers are using it right now.
The skill was added a while ago so it's not in the preview swagger.
Now our team wants to GA the feature so it should live in both the GA API and the preview API. 
Given the situation we are at, should I remove or keep the 2020-06-30 change?

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (shuyangmsft) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6970]: There are optional parameters. If customers don't provide any value, our backend service will get a null, which is an equivalence of "false".

File: specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json
Comment (heaths) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: /cc @johanste @brjohnstmsft

Comment (shuyangmsft) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: hi team, can I get any update on this?

Comment (brjohnstmsft) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: FYI @bleroy since this is about Skillsets

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (bleroy) [inline specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6974]: The wording "sensitive to accent" is very weird. Is it going to be sensitive to my French accent? :D Maybe something more like "considers accented and unaccented letters such as 'é' and 'e' to be identical"

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (bleroy) [inline specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6983]: Typo: missing "can" in "It be". Is it also missing an apostrophe on "aliases"?

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (bleroy) [inline specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6999]: Can we say "CustomEntityAlias" instead of just "complex objects"?

File: specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json
Comment (bleroy) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: This should be done consistently with our new story for skill versioning that @Careyjmac is building (provided timing allows it, but I'd still like to have her input for this specific case), but it should also be ruled by the general rules about breaking changes for existing APIs since this for all practical purposes seems like it's GA. I agree with @heaths' original comment that the GA version should be left alone and only the preview version should have the changes.

**Review by bleroy — CHANGES_REQUESTED** [review]

File: specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json
Comment (Careyjmac) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: Our skill story with regards to preview/GA is currently in a weird state.  Technically in the API on the backend, skills are version neutral and thus whenever we add a new skill or change an existing one, that change occurs across all API versions that support skillsets. We like to call skills "preview" when they are new by adding a preview disclaimer to the documentation and not adding SDK support for them until we decide to "GA" them but this is handwave-y because they are already technically exposed in GA API versions.

We understand this is probably not a very good practice and we are looking into ways to make the preview experience better for skills as part of our skill versioning design, which is scheduled to be implemented in Cobalt. For now, because we have decided to "GA" the CustomEntityLookupSkill via our handwave-y method of removing the preview disclaimer from the docs and adding it the the GA SDK swagger, Shuyang's change here makes sense.

**Review by bleroy — APPROVED** [review]

File: specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json
Comment (heaths) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: @Careyjmac please also consider the impact and versioning story of the SDK. To support this, the SDKs have to be regenerated, which first includes updating the commits to the swagger files. But since new APIs were added, strict semver rules dictate you now bump the minor version of the SDK packages. If skills (or any other models or methods) are added with some frequency, the package versions will iterate as frequently. This might become daunting for customers to know if they should upgrade when considering bug fixes will bump the patch version.

In such cases, bumping only the patch version for new models in the SDKs may make more sense, but is something the service and SDK teams should consider together.

Comment (Careyjmac) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: @Heaths of course.  We are still getting used to navigating this new world where the SDK team is so involved and all of the new breaking change requirements.  We are definitely considering the SDK implications with the new skill versioning discussions.  But since those discussions are ongoing and the work won't be done until Cobalt, I think taking a minor version bump for a change such as this one is acceptable for now.

Comment (heaths) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: We're going to GA a minor version bump in a few weeks. @AlexGhiondea would adding a model require going through at least a single preview, or would this be minor enough (no pun intended) to include in our upcoming GA?

**Review by tjprescott — APPROVED** [review]

File: specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json
Comment (heaths) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: @johanste how should this be handled going forward if new models are added to GAs? Isn't there roundtripping concerns here as well?

Comment (johanste) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: Yes, there are some round-tripping concerns here. A client that knows nothing about a new skill type can easily end up stripping out properties it did not know about on update (assuming PUT does a replace rather than a merge-patch like update). I believe that I and @JeffreyRichter can be of assistance when it comes to determining (or at least reviewing) what a good versioning model for your service APIs should be.
Likewise, as mentioned above,

Comment (Careyjmac) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30/searchservice.json:6946]: + @jennifermarsman since I know she was talking about revisting this design in the coming weeks.  We are hoping to finalize the design in Iron with the intent of implementing it in Cobalt, and we would definitely value any feedback from the SDK folks as we do so, so perhaps we should loop ya'll into our thread we have on the topic.

File: specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json
Comment (heaths) [reply specification/search/data-plane/Azure.Search/preview/2020-06-30-Preview/searchservice.json:6947]: But I bet you send them back with explicit nulls, which breaks the deserializers. The point of saving bandwidth was that you NOT send back values with explicit nulls. Putting x-nullable in the swagger - which isn't transmitted with every request - saves no bandwidth.

Last I heard from @brjohnstmsft it was the OData returning values with explicit nulls. If you can't turn that off, you need to mark any optional parameters with x-nullable: true.

**Review by heaths — CHANGES_REQUESTED** [review]
Need proper x-nullable attribution. Lack of these and returning explicit nulls has been breaking the SDK deserializers that respect this attribution and often depend on it. So far, we've been playing whack-a-mole with customer-reported bugs trying to fix all the missing locations up. I found many with existing skills.

**Review by heaths — APPROVED** [review]

**Comment by tjprescott** [comment]
@shuyangmsft is this ready to merge?

**Comment by shuyangmsft** [comment]
> @shuyangmsft is this ready to merge?

**Comment by shuyangmsft** [comment]
> @shuyangmsft is this ready to merge?

Hi Travis, our backend change is in pipeline. I'll let you know when it's deployed to prod. Thanks!

**Comment by heaths** [comment]
Any ETA? We have an SDK release coming up very soon and still need to regenerate source based on this change.

**Comment by shuyangmsft** [comment]
> Any ETA? We have an SDK release coming up very soon and still need to regenerate source based on this change.

Based on our pipeline status, we definitely can't make this change to prod this week. Next week should work. I'll keep you updated.

**Comment by heaths** [comment]
Please hold off on merging this. We won't be able to generate and test this sufficiently for the upcoming release and have decided not to include it. It will have to be included in the next beta. This can be merged after we code freeze next week.

**Review by tjprescott — APPROVED** [review]

**Pull request merged**
